babashka: Add -H:PageSize=64K build option on aarch64

diegobfernandez opened the pull request:
## Description of changes

Babashka builds upstream with `-H:PageSize=64K` for `aarch64`. This PR updates things accordingly.

The change enables the package to be used with [all aarch64 kernel page sizes (4K, 16K and 64K) with the downside of a bit of wasted memory](https://github.com/oracle/graal/pull/5605#issuecomment-1364006130).

Here is the PR from babashka that I based the changes on: https://github.com/babashka/babashka/pull/1606

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux (4K and 16K kernel page size)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

diegobfernandez committed:
babashka: Add -H:PageSize=64K build option on aarch64

Makes babashka compatible with all aarch64 page sizes, 4K, 16K and 64K.
Upstream changes: https://github.com/babashka/babashka/pull/1606

thiagokokada commented:
What are the implications of this change for usage in non-Asahi systems?

thiagokokada reviewed (changes requested):
Please fix your commits accordingly to https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions.

diegobfernandez commented:
Thanks for the quick reply @thiagokokada. I updated the "Description of changes" to be more explicit on the reasons and effects of the change.
There should be no impact on non-Asahi. I also removed the mention to asahi as this issue applies to any non-4K kernels.

Also fixing the commits as requested.

thiagokokada reviewed (changes requested):
I have some concerns about the current implementation:
- How compatible is `nixpkgs` with Asahi in general? Are most binaries (especially the ones build from source with `stdenv`, e.g.: most programs that link to `glibc`) compatible with Asahi? Otherwise I don't see a good motivation to fix one package only, it seems that if you want to run `babashka` in Asahi it would be better to run the static binary from upstream
- Does this change brings any significant performance impact? Asahi is a really small amount of users and making the performance worse for every other user there (especially considering that most ARM systems are underpowered) seems unwisely
- If both questions above have acceptable answers, them I would prefer to fix this in `buildGraalvmNativeImage` instead. Otherwise all other Native Image binaries (e.g.: `clojure-lsp`) would still be broken in Asahi (BTW, can someone test if other Native Image binaries are broken in Asahi?)

bennyandresen commented:
Throwing my comment into the ring: I think having an explicit `PageSize` value set in a `buildGraalvmNativeImage` would be a good move, instead of relying on the default behavior of `PageSize=0`.

thiagokokada commented:
> Throwing my comment into the ring: I think having an explicit `PageSize` value set in a `buildGraalvmNativeImage` would be a good move, instead of relying on the default behavior of `PageSize=0`.

Given everything that @diegobfernandez I would be fine having this set directly in `buildGraalvmNativeImage` too, however only in `aarch64-linux` systems because I think it is better to let macOS binaries be build with their default `PageSize` (that will be 64k in this case, since I imagine this is the the default in `aarch64-darwin`).

diegobfernandez commented:
Got it. I will close this one and reference it in a new PR with the changes in `buildGraalvmNativeImage`. The changes are ready I just didn't had the chance to test all packages that rely on it (so far I tested babashka, clojure-lsp and cljfmt).
